File: easy/001_Two_Sum/two_sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        a = [(nums[i], i) for i in xrange(0, len(nums))]
        a.sort(key=lambda x:x[0])
        i, j = 0, len(nums)-1

        while i < j:
            total = a[i][0] + a[j][0]
            if total == target:
                return [a[i][1], a[j][1]]
            elif total < target:
                i += 1
            else:
                j -= 1

        return [-1, -1]

solution = Solution()

a0 = [1, 2, 3, 50, 52, 55]
print solution.twoSum(a0, 5)
print solution.twoSum(a0, 3)
print solution.twoSum(a0, 55)

a1 = [3, 2, 4]
print solution.twoSum(a1, 6)

File: easy/001_Two_Sum/twoSum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	tbl := make(map[int][]int, len(nums))
	for i, num := range nums {
		_, ok := tbl[num]
		if !ok {
			tbl[num] = []int{i}
		} else {
			tbl[num] = append(tbl[num], i)
		}
	}

	for i, num := range nums {
		wanted := target - num
		a, ok := tbl[wanted]
		if !ok {
			continue
		}
		for _, j := range a {
			if j != i {
				return []int{i, j}
			}
		}
	}
	return nil
}

func main() {
	nums := []int{1, 5, 7, 9}
	target := 12
	fmt.Printf("twoSum %v\n", twoSum(nums, target))
}
